### DISCUSSION TITLE
Problem with CytoNorm Prep

### DISCUSSION THREAD

**Discussion by MBakker06** [ImmuneDynamics/Spectre#165] [Help] [answered]
Hi, I'm running into an issue when trying to run the following code during batch alignment:

`align.model <- prep.cytonorm(dat = ref.dat,
                             cellular.cols = c(cellular.cols, paste0(cellular.cols, crs.append)),
                             cluster.cols = paste0(cluster.cols, crs.append),
                             batch.col = batch.col,
                             dir = fine.dir,
                             xdim = 14,
                             ydim = 14,
                             meta.k = 8)`

It gets stuck during step 2/4 and then I receive the following error:

Step 1/4 - Splitting files for use with original FlowSOM function
Step 2/4 - Running FlowSOM
Reading 1.fcs
**Error in if (sampleWithReplacement & (nrow(f) < cFile)) { : 
  argument is of length zero**

I followed the Batch alignment code from the general discovery workflow:
[https://github.com/ImmuneDynamics/Spectre/blob/master/workflows/Classic%20workflows/Spectre%20general%20discovery%20workflow/2%20-%20batch%20alignment.R](url)

Does anyone have an idea how to resolve this issue, any help is much appreciated! :)

**Comment by tomashhurst** [comment]
Hi @MBakker06 , thanks for reaching out! If you run the code here on the workflow page (https://wiki.centenary.org.au/display/SPECTRE/Discovery+workflow+with+batch+alignment+using+CytoNorm), do you get a successful result?

**Reply by MBakker06** [reply]
Hi, thank you for your response! I have tried the code from the workflow page. However, I get the exact same error. I follow the code by the letter, but cannot get it to work. One difference is that my data is imported from .fcs and not .csv. The strange thing is that the coarse batch alignment does work, but the fine alignment doesn't. Is there anything else I can try? Thank you very much in advance.

**Reply by MBakker06** [reply]
We tried the code on that workflow page with the demo data as well and still run into the same error.

**Reply by ghar1821** [reply]
Hi @MBakker06,

Can you try the workflow in the link @tomashhurst provided using the dataset (in the data and metadata folder) from this github link (https://github.com/ImmuneDynamics/Spectre/tree/master/workflows/Alignment%20workflows)? That is the dataset that is used in the workflow. I just tested at my end and everything seem to be working fine. If it doesn't work at your end, can you post the error message so we can help you debug the problem better?

**Reply by MBakker06** [reply]
Hi, I have followed the code and used the data from that github link. I still receive the following error:

Working directory is 'C:/XXX/Output_Spectre/Output 2 - alignment/1 - ref pre-alignment'
Step 1/3. Mapping data
Step 2/3. Merging data
Step 3/3. Returning data
Step 1/4 - Splitting files for use with original FlowSOM function
Step 2/4 - Running FlowSOM
Reading A.fcs
Error in if (sampleWithReplacement & (nrow(f) < cFile)) { : 
  argument is of length zero


All packages are updated. I'm running the newest versions of Rstudio and R. I cannot find the problem. A coworker tried the same code and data on another computer, but he received the same error message. Any help and ideas are very welcome! I really want to figure out what is going on.

**Reply by ghar1821** [reply]
Can you run `sessionInfo()` and post the content? This is to see what versions of what packages you are using.

**Reply by MBakker06** [reply]
> sessionInfo()
R version 4.3.0 (2023-04-21 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=English_Netherlands.utf8  LC_CTYPE=English_Netherlands.utf8    LC_MONETARY=English_Netherlands.utf8
[4] LC_NUMERIC=C                         LC_TIME=English_Netherlands.utf8    

time zone: Europe/Amsterdam
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] Spectre_1.0.0     compiler_4.3.0    tools_4.3.0       rstudioapi_0.14   data.table_1.14.8

**Reply by ghar1821** [reply]
Ah I'm hoping to find the version of FlowSOM and Cytonorm you have installed. Can you find that and paste them here? You can use `packageVersion("flowsom")` function (replace flowsom with cytonorm for cytonorm). Also, did u install Spectre from our master or development branch?

**Reply by MBakker06** [reply]
I installed Spectre from the master branch.
FlowSOM v2.8.0
CytoNorm v0.0.17 
(CytoNorm was installed with `install_github('saeyslab/CytoNorm')` using devtools

**Reply by ghar1821** [reply]
Can you try re-installing Spectre using the development branch? You can use `install_github(immunedynamics/Spectre, ref="development")`. We may have had some issues with the alignment functions in the past which may have been resolved in our development branch but not yet pushed to the master branch.

**Comment by mortadelo93** [comment]
I have the same issue.
Did anything work?

**Reply by ghar1821** [reply]
we recently deployed a new version. Can you try re-installing and see if it solves the problem?

**Reply by mortadelo93** [reply]
Hi
Thanks for the replay.

Im afraid the error still persist.

I uninstalled Spectre and installed again using
devtools::install_github("immunedynamics/spectre")

I run again the workflow and nothing changed.
The error is the same as @MBakker06

>Step 1/4 - Splitting files for use with original FlowSOM function
Step 2/4 - Running FlowSOM
Reading batch1.fcs
Error in if (sampleWithReplacement & (nrow(f) < cFile)) { : 
  argument is of length zero

> package.version('Spectre')
[1] "1.1.0"
> package.version('FlowSOM')
[1] "2.10.0"
> package.version('CytoNorm')
[1] "2.0.0"
> session_info()
─ Session info ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Sonoma 14.1
 system   aarch64, darwin20
 ui       RStudio

**Comment by SamGG** [comment]
> Error in if (sampleWithReplacement & (nrow(f) < cFile)) { :
> argument is of length zero

& should be replaced by &&

**Reply by mortadelo93** [reply]
Should I modify original code using trace() or something like that?

**Reply by SamGG** [reply]
This line of code belongs to the FlowSOM package... I referenced the issue in https://github.com/saeyslab/FlowSOM/issues/18.
The question still why either sampleWithReplacement or nrow(f) or cFile is of length zero...
I don't know how to modify code using trace().
What I would do is to fork the github repo, modify the code changing & by && and adding a `browser()` command, and install the modified version FlowSOM. But only if you feel confident.
Best.

**Reply by mortadelo93** [reply]
I guess I could have a try.
Hoping developers fix it too.
Thanks for the info @SamGG

**Reply by mortadelo93** [reply]
Just an update if someone has the same issue.
The error as @SamGG said is in FlowSOM, specifically in FlowSOM::AggregateFlowFrames() function.

What I do to fix it is to run trace(FlowSOM::AggregateFlowFrames, edit = T) and replace the & by &&.
I have to do this every new session but I could not modify the original version of the package.

Thanks again

**Comment by tomashhurst** [comment, marked as answer]
The `prep.cytonorm` issues should be fixed in the beta of v1.2 which you can test out with the following install code.

```
if(!require('remotes')) {install.packages('remotes')} # Installs the package 'remotes'
remotes::install_github(repo = "immunedynamics/spectre", ref = 'v1.2.0-beta') # Install the Spectre package
```

Don't forget the `ref = 'v1.2.0-beta'`

We are hoping to have v1.2 out as the main release next week.
